// @mui/x-data-grid-generator/modern/renderer/renderEditRating.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import Box from '@mui/material/Box';
import Rating from '@mui/material/Rating';
import { useGridApiContext } from '@mui/x-data-grid-premium';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
function EditRating(props) {
  const {
    id,
    value,
    field
  } = props;
  const apiRef = useGridApiContext();
  const handleChange = async event => {
    await apiRef.current.setEditCellValue({
      id,
      field,
      value: Number(event.target.value)
    }, event);
    apiRef.current.stopCellEditMode({
      id,
      field
    });
  };
  const handleRef = element => {
    if (element) {
      element.querySelector(`input[value="${value}"]`).focus();
    }
  };
  return /*#__PURE__*/_jsxs(Box, {
    sx: {
      display: 'flex',
      alignItems: 'center',
      lineHeight: '24px',
      color: 'text.secondary',
      mr: 1
    },
    children: [/*#__PURE__*/_jsx(Rating, {
      ref: handleRef,
      name: "rating",
      value: Number(value),
      precision: 1,
      onChange: handleChange,
      sx: {
        mr: 1
      }
    }), Number(value)]
  });
}
export function renderEditRating(params) {
  return /*#__PURE__*/_jsx(EditRating, _extends({}, params));
}
