// @mui/x-data-grid-generator/legacy/renderer/renderEditCurrency.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _extends from "@babel/runtime/helpers/esm/extends";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import * as React from 'react';
import { useGridApiContext } from '@mui/x-data-grid-premium';
import Autocomplete, { autocompleteClasses } from '@mui/material/Autocomplete';
import InputBase from '@mui/material/InputBase';
import Box from '@mui/material/Box';
import { styled } from '@mui/material/styles';
import { CURRENCY_OPTIONS } from '../services/static-data';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var StyledAutocomplete = styled(Autocomplete)(function (_ref) {
  var theme = _ref.theme;
  return _defineProperty({
    height: '100%'
  }, "& .".concat(autocompleteClasses.inputRoot), _extends({}, theme.typography.body2, {
    padding: '1px 0',
    height: '100%',
    '& input': {
      padding: '0 16px',
      height: '100%'
    }
  }));
});
function EditCurrency(props) {
  var id = props.id,
    value = props.value,
    field = props.field;
  var apiRef = useGridApiContext();
  var handleChange = React.useCallback( /*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(event, newValue) {
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return apiRef.current.setEditCellValue({
              id: id,
              field: field,
              value: newValue.toUpperCase()
            }, event);
          case 2:
            apiRef.current.stopCellEditMode({
              id: id,
              field: field
            });
          case 3:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function (_x, _x2) {
      return _ref3.apply(this, arguments);
    };
  }(), [apiRef, field, id]);
  return /*#__PURE__*/_jsx(StyledAutocomplete, {
    value: value,
    onChange: handleChange,
    options: CURRENCY_OPTIONS,
    autoHighlight: true,
    fullWidth: true,
    open: true,
    disableClearable: true,
    renderOption: function renderOption(optionProps, option) {
      return /*#__PURE__*/_jsxs(Box, _extends({
        component: "li",
        sx: {
          '& > img': {
            mr: 1.5,
            flexShrink: 0
          }
        }
      }, optionProps, {
        children: [/*#__PURE__*/_jsx("img", {
          loading: "lazy",
          width: "20",
          src: "https://flagcdn.com/w20/".concat(option.slice(0, -1).toLowerCase(), ".png"),
          srcSet: "https://flagcdn.com/w40/".concat(option.slice(0, -1).toLowerCase(), ".png 2x"),
          alt: ""
        }), option]
      }));
    },
    renderInput: function renderInput(params) {
      return /*#__PURE__*/_jsx(InputBase, _extends({
        autoFocus: true,
        fullWidth: true,
        id: params.id,
        inputProps: _extends({}, params.inputProps, {
          autoComplete: 'new-password' // disable autocomplete and autofill
        })
      }, params.InputProps));
    }
  });
}
export function renderEditCurrency(params) {
  return /*#__PURE__*/_jsx(EditCurrency, _extends({}, params));
}
