// @mui/x-data-grid-generator/legacy/renderer/renderEditProgress.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as React from 'react';
import clsx from 'clsx';
import { useGridApiContext } from '@mui/x-data-grid-premium';
import Slider, { sliderClasses } from '@mui/material/Slider';
import Tooltip from '@mui/material/Tooltip';
import { unstable_debounce as debounce } from '@mui/utils';
import { alpha, styled } from '@mui/material/styles';
import { jsx as _jsx } from "react/jsx-runtime";
var StyledSlider = styled(Slider)(function (_ref) {
  var theme = _ref.theme;
  return _defineProperty(_defineProperty(_defineProperty({
    display: 'flex',
    height: '100%',
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 0,
    borderRadius: 0
  }, "& .".concat(sliderClasses.rail), {
    height: '100%',
    backgroundColor: 'transparent'
  }), "& .".concat(sliderClasses.track), {
    height: '100%',
    transition: theme.transitions.create('background-color', {
      duration: theme.transitions.duration.shorter
    }),
    '&.low': {
      backgroundColor: '#f44336'
    },
    '&.medium': {
      backgroundColor: '#efbb5aa3'
    },
    '&.high': {
      backgroundColor: '#088208a3'
    }
  }), "& .".concat(sliderClasses.thumb), {
    height: '100%',
    width: 5,
    borderRadius: 0,
    marginTop: 0,
    backgroundColor: alpha('#000000', 0.2)
  });
});
function ValueLabelComponent(props) {
  var children = props.children,
    open = props.open,
    value = props.value;
  return /*#__PURE__*/_jsx(Tooltip, {
    open: open,
    enterTouchDelay: 0,
    placement: "top",
    title: value,
    children: children
  });
}
function EditProgress(props) {
  var id = props.id,
    value = props.value,
    field = props.field;
  var _React$useState = React.useState(Number(value)),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    valueState = _React$useState2[0],
    setValueState = _React$useState2[1];
  var apiRef = useGridApiContext();
  var updateCellEditProps = React.useCallback(function (newValue) {
    apiRef.current.setEditCellValue({
      id: id,
      field: field,
      value: newValue
    });
  }, [apiRef, field, id]);
  var debouncedUpdateCellEditProps = React.useMemo(function () {
    return debounce(updateCellEditProps, 60);
  }, [updateCellEditProps]);
  var handleChange = function handleChange(event, newValue) {
    setValueState(newValue);
    debouncedUpdateCellEditProps(newValue);
  };
  React.useEffect(function () {
    setValueState(Number(value));
  }, [value]);
  var handleRef = function handleRef(element) {
    if (element) {
      element.querySelector('[type="range"]').focus();
    }
  };
  return /*#__PURE__*/_jsx(StyledSlider, {
    ref: handleRef,
    classes: {
      track: clsx(valueState < 0.3 && "low", valueState >= 0.3 && valueState <= 0.7 && "medium", valueState > 0.7 && "high")
    },
    value: valueState,
    max: 1,
    step: 0.00001,
    onChange: handleChange,
    components: {
      ValueLabel: ValueLabelComponent
    },
    valueLabelDisplay: "auto",
    valueLabelFormat: function valueLabelFormat(newValue) {
      return "".concat((newValue * 100).toLocaleString(), " %");
    }
  });
}
export function renderEditProgress(params) {
  return /*#__PURE__*/_jsx(EditProgress, _extends({}, params));
}
